flang/test/Lower/OpenMP/omp-sections.f90

```diff
@@ -0,0 +1,108 @@
+!FIRDialect: %8 = fir.convert %7 : (i32) -> f32
+!FIRDialect: fir.store %8 to %4 : !fir.ref<f32>
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.section {
+!FIRDialect: {{.*}} = fir.load {{.*}} : !fir.ref<i32>
+!FIRDialect: {{.*}} = arith.constant 1 : i32
+!FIRDialect: {{.*}} = arith.addi {{.*}} : i32
+!FIRDialect: fir.store {{.*}} to {{.*}} : !fir.ref<i32>
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.section {
+!FIRDialect: {{.*}} = fir.load {{.*}} : !fir.ref<f32>
+!FIRDialect: {{.*}} = arith.constant 7.000000e+00 : f32
+!FIRDialect: {{.*}} = arith.subf {{.*}} : f32
+!FIRDialect: fir.store {{.*}} to {{.*}} : !fir.ref<f32>
+!FIRDialect: %7 = fir.load %2 : !fir.ref<i32>
+!FIRDialect: %8 = fir.convert %7 : (i32) -> f32
+!FIRDialect: %9 = fir.load %4 : !fir.ref<f32>
+!FIRDialect: %10 = arith.mulf %8, %9 : f32
+!FIRDialect: %11 = fir.convert %10 : (f32) -> i32
+!FIRDialect: fir.store %11 to %2 : !fir.ref<i32>
+!FIRDialect: %12 = fir.load %2 : !fir.ref<i32>
+!FIRDialect: %13 = fir.convert %12 : (i32) -> f32
+!FIRDialect: %14 = fir.load %4 : !fir.ref<f32>
+!FIRDialect: %15 = arith.subf %13, %14 : f32
+!FIRDialect: %16 = fir.convert %15 : (f32) -> i32
+!FIRDialect: fir.store %16 to %3 : !fir.ref<i32>
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.sections nowait {
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: return
+!FIRDialect: }
+
+program sample
+    use omp_lib
+    integer :: count = 0, double_count = 1 
+    !$omp sections private (count, double_count) allocate(omp_default_mem_alloc: eta)  
+        !$omp section
+            count = 1 + 4
+            eta = count * double_count
+        !$omp section
+            double_count = double_count + 1
+        !$omp section
+            eta = eta - 7
+            count = count * eta
+            double_count = count - eta
+    !$omp end sections
+
+    !$omp sections
+    !$omp end sections nowait
+end program sample
+
+!FIRDialect: func @_QPfirstprivate(%arg0: !fir.ref<f32>) {
+!FIRDialect: %[[ALPHA:.*]] = fir.alloca f32 {bindc_name = "alpha", pinned, uniq_name = "_QFfirstprivateEalpha"}
+!FIRDialect: %[[ALPHA_STORE:.*]] = fir.load %arg0 : !fir.ref<f32>
+!FIRDialect: fir.store %[[ALPHA_STORE]] to %[[ALPHA]] : !fir.ref<f32>
+!FIRDialect: omp.sections {
+!FIRDialect: omp.section  {
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.sections {
+!FIRDialect: omp.section  {
+!FIRDialect: %2 = fir.load %arg0 : !fir.ref<f32>
+!FIRDialect: %cst = arith.constant 5.000000e+00 : f32
+!FIRDialect: %3 = arith.mulf %2, %cst : f32
+!FIRDialect: fir.store %3 to %arg0 : !fir.ref<f32>
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: omp.terminator
+!FIRDialect: }
+!FIRDialect: return
+!FIRDialect: }
+
+subroutine firstprivate(alpha)
+    real :: alpha 
+    !$omp sections firstprivate(alpha)
+    !$omp end sections
+
+    !$omp sections
```

Comment:
@kiranchandramohan @shraiysh In case no `!$omp section` is present, I am observing a default `!$omp section` in the parse tree, and accordingly a default `omp.section` in the lowering. Is this expected behavior?

![default_block](https://user-images.githubusercontent.com/42909663/145588017-5336f7b4-123e-454f-8022-84a887d4af42.png)

Reply:
Yes, that is okay. The first `omp section` is optional but semantically it is always there.